File: backend/fx.py
"""Foreign-currency → ILS for Leumi Visa exports.

When the bank provides column 3 (charge_amount in ILS), that value is always used.
Live FX rates are fetched only for pending charges with no bank ILS amount yet.
"""
from __future__ import annotations

import re
from datetime import date
from typing import Optional

from fx_rates import get_rate_to_ils

_HEBREW = re.compile(r"[\u0590-\u05FF]")
_COUNTRY_SUFFIX = re.compile(r"\s([A-Z]{2})\s*$")
_USD_MERCHANT = re.compile(
    r"OPENAI|CURSOR|AWS\s|APPLE\.COM|MIDJOURNEY|AIRBNB|GOOGLE|GITHUB|MICROSOFT|"
    r"NETFLIX|SPOTIFY|TRIP\.COM|AIRALO|OMIO|BOUNCE|FILTERLY|SPECIALSCOMEDY|"
    r"yollacalls|CONCERTPLACE|BANDCAMP|economybookings|TradeInn|Farfetch|"
    r"HOTEL & RESORTS|DRIVALIA|SKIS ROSSIGNOL|EL AL \d",
    re.IGNORECASE,
)
_EUR_MERCHANT = re.compile(
    r"\b(SRL|SPA|OOD|EOOD|GMBH|TERMINI|ROMA|MILANO|AUTOGRILL|UBR\*|AVIS|ZARA|COS|"
    r"UNIQLO|POLENE|VIvat|PAUL AIRPORT|DEDEM SPA|PIERLUIGI|SUMUP|MANGO ROMA)\b",
    re.IGNORECASE,
)
_BGN_MERCHANT = re.compile(r"\b(EOOD|OOD|BALGARIYA|BURGAS|BULGARIA|AMREST KOFI)\b", re.IGNORECASE)
_ILS_MERCHANT = re.compile(r"\bBIT\b", re.IGNORECASE)


def _round_money(value: float) -> float:
    return round(value * 100) / 100


def _has_hebrew(text: str) -> bool:
    return bool(_HEBREW.search(text))


def infer_currency_from_ratio(amount: float, charge: float) -> Optional[str]:
    if amount <= 0 or charge <= 0:
        return None
    ratio = charge / amount
    if 3.2 <= ratio <= 4.05:
        return "USD"
    if 3.85 <= ratio <= 4.55:
        return "EUR"
    if 1.85 <= ratio <= 2.25:
        return "BGN"
    if 4.5 <= ratio <= 5.0:
        return "GBP"
    if 0.85 <= ratio <= 1.05:
        return "PLN"
    if 0.005 <= ratio <= 0.02:
        return "AMD"
    if 0.98 <= ratio <= 1.02:
        return "ILS"
    return None


def detect_currency(
    merchant: str,
    amount: float,
    charge: Optional[float],
    explicit_currency: Optional[str] = None,
) -> str:
    if explicit_currency and explicit_currency != "ILS":
        return explicit_currency

    if _has_hebrew(merchant):
        return "ILS"

    if _ILS_MERCHANT.search(merchant):
        return "ILS"

    suffix = _COUNTRY_SUFFIX.search(merchant.strip())
    if suffix and suffix.group(1) == "IL":
        return "ILS"

    if charge is not None and charge > 0 and amount > 0:
        inferred = infer_currency_from_ratio(amount, charge)
        if inferred:
            return inferred

    if _USD_MERCHANT.search(merchant):
        return "USD"

    if suffix:
        cc = suffix.group(1)
        if cc == "IL":
            return "ILS"
        if cc == "US":
            return "USD"
        if cc == "PL":
            return "PLN"
        if cc in ("DE", "FR", "IT", "ES", "NL", "AT", "BE", "PT", "GR"):
            return "EUR"
        if cc == "GB":
            return "GBP"
        if cc == "BG":
            return "BGN"
        if cc == "AM":
            return "AMD"

    if _BGN_MERCHANT.search(merchant):
        return "BGN"
    if _EUR_MERCHANT.search(merchant):
        return "EUR"

    if re.search(r"[A-Za-z]{3,}", merchant) and not _has_hebrew(merchant):
        return "USD"

    return "ILS"


REFUND_TYPE_MARKERS = ("זיכוי", "השבת מכירה")
_INSTALLMENT = re.compile(r"תשלום\s+(\d+)\s+מתוך\s+(\d+)")
_BILLING_ADJUSTMENT = "שינוי מועד חיוב"
_CARD_FEE_TYPE = "דמי חבר"
_INTEREST_MERCHANT = "* ריבית *"


def parse_installment(notes: Optional[str]) -> Optional[tuple[int, int]]:
    if not notes:
        return None
    match = _INSTALLMENT.search(notes)
    if not match:
        return None
    return int(match.group(1)), int(match.group(2))


def should_skip_installment_row(
    charge_raw,
    notes: Optional[str],
    *,
    transaction_type_he: Optional[str] = None,
) -> bool:
    """Credit installments before the bank posts an ILS slice — skip, do not estimate the full amount."""
    installment = parse_installment(notes)
    if installment is None:
        return False
    current, _ = installment
    if current == 0:
        return True
    return is_pending_charge(
        charge_raw,
        notes,
        transaction_type_he=transaction_type_he,
    )


def is_billing_cycle_adjustment(merchant: str) -> bool:
    return _BILLING_ADJUSTMENT in merchant


def is_card_membership_fee(merchant: str, transaction_type_he: Optional[str] = None) -> bool:
    if transaction_type_he == _CARD_FEE_TYPE:
        return True
    return "דמי כרטיס" in merchant


def is_interest_row(merchant: str) -> bool:
    return _INTEREST_MERCHANT in merchant


def should_skip_non_spend_row(
    charge_raw,
    merchant: str,
    notes: Optional[str],
    *,
    transaction_type_he: Optional[str] = None,
) -> bool:
    """Rows that must not appear in spend totals or transaction lists."""
    if should_skip_installment_row(charge_raw, notes, transaction_type_he=transaction_type_he):
        return True
    if is_billing_cycle_adjustment(merchant):
        return True
    if is_card_membership_fee(merchant, transaction_type_he) and is_pending_charge(
        charge_raw,
        notes,
        transaction_type_he=transaction_type_he,
    ):
        return True
    return False


def transaction_snapshot_key(tx) -> tuple:
    """Dedupe key — installment payments on the same day stay distinct."""
    raw_date = getattr(tx, "date", None)
    if isinstance(raw_date, date):
        tx_date = raw_date.isoformat()
    else:
        tx_date = str(raw_date)[:10]
    merchant = getattr(tx, "merchant_he", "")
    notes = getattr(tx, "notes", None)
    installment = parse_installment(notes)
    if installment:
        current, total = installment
        return (tx_date, merchant, "inst", current, total)
    charge = round(float(getattr(tx, "charge_amount", 0) or 0), 2)
    note_key = notes or ""
    return (tx_date, merchant, "tx", charge, note_key)


def pick_better_transaction(a, b):
    """When the same purchase appears in multiple exports, keep the billed snapshot."""
    a_est = bool(getattr(a, "charge_estimated", False))
    b_est = bool(getattr(b, "charge_estimated", False))
    if a_est != b_est:
        return b if a_est else a

    a_inst = parse_installment(getattr(a, "notes", None))
    b_inst = parse_installment(getattr(b, "notes", None))
    if a_inst and b_inst:
        if a_inst[0] != b_inst[0]:
            return a if a_inst[0] > b_inst[0] else b

    if a.charge_amount != b.charge_amount:
        return a if a.charge_amount < b.charge_amount else b

    return b


def dedupe_transaction_snapshots(transactions: list) -> list:
    best: dict[tuple, object] = {}
    order: list[tuple] = []
    for tx in transactions:
        key = transaction_snapshot_key(tx)
        if key not in best:
            order.append(key)
            best[key] = tx
        else:
            best[key] = pick_better_transaction(best[key], tx)
    return [best[key] for key in order]


def is_refund_transaction(
    transaction_type_he: Optional[str],
    amount: float,
    charge_raw,
) -> bool:
    if transaction_type_he and any(m in transaction_type_he for m in REFUND_TYPE_MARKERS):
        return True
    if amount < 0:
        return True
    if charge_raw is not None:
        try:
            return float(charge_raw) < 0
        except (TypeError, ValueError):
            pass
    return False


def _tx_date_value(tx) -> date:
    raw = getattr(tx, "date", None)
    if isinstance(raw, date):
        return raw
    if isinstance(raw, str):
        return date.fromisoformat(raw[:10])
    raise TypeError(f"Unsupported transaction date: {raw!r}")


def find_matching_original_charge(
    merchant: str,
    foreign_amount: float,
    refund_date: date,
    transactions: list,
    *,
    skip_index: Optional[int] = None,
):
    """Find a prior purchase with the same merchant and foreign amount."""
    if foreign_amount <= 0:
        return None
    best = None
    best_date: Optional[date] = None
    for i, tx in enumerate(transactions):
        if skip_index is not None and i == skip_index:
            continue
        if getattr(tx, "merchant_he", "") != merchant:
            continue
        tx_amount = float(getattr(tx, "amount", 0))
        tx_charge = float(getattr(tx, "charge_amount", 0))
        tx_type = getattr(tx, "transaction_type_he", None)
        if is_refund_transaction(tx_type, tx_amount, tx_charge if tx_charge != 0 else None):
            continue
        if tx_charge <= 0:
            continue
        if abs(abs(tx_amount) - foreign_amount) > 0.02:
            continue
        tx_date = _tx_date_value(tx)
        if tx_date > refund_date:
            continue
        if best is None or tx_date > best_date:
            best = tx
            best_date = tx_date
    return best


def resolve_refund_ils(
    amount: float,
    charge_raw,
    merchant: str,
    tx_date: Optional[date] = None,
    explicit_currency: Optional[str] = None,
    transaction_type_he: Optional[str] = None,
    transactions_for_matching: Optional[list] = None,
    skip_index: Optional[int] = None,
) -> tuple[float, Optional[str], bool]:
    """Match refund to original charge when possible; otherwise convert via FX."""
    charge: Optional[float]
    if charge_raw is None:
        charge = None
    else:
        try:
            charge = float(charge_raw)
        except (TypeError, ValueError):
            charge = None

    foreign_amount = abs(amount) if amount != 0 else (abs(charge) if charge else 0.0)
    refund_date = tx_date or date.today()

    if transactions_for_matching:
        match = find_matching_original_charge(
            merchant,
            foreign_amount,
            refund_date,
            transactions_for_matching,
            skip_index=skip_index,
        )
        if match is not None:
            original_currency = getattr(match, "original_currency", None) or "ILS"
            estimated = bool(getattr(match, "charge_estimated", False))
            return _round_money(-abs(float(match.charge_amount))), original_currency, estimated

    if charge is not None and charge != 0 and foreign_amount > 0:
        ils_magnitude = abs(charge)
        if abs(ils_magnitude - foreign_amount) > 0.02:
            return _round_money(-ils_magnitude), "ILS", False

    if explicit_currency and explicit_currency != "ILS":
        rate = get_rate_to_ils(explicit_currency, refund_date)
        return _round_money(-foreign_amount * rate), explicit_currency, True

    base = abs(charge) if charge is not None and charge != 0 else foreign_amount
    return _round_money(-base), "ILS", False


def is_pending_charge(
    charge_raw,
    notes: Optional[str],
    *,
    transaction_type_he: Optional[str] = None,
    amount: Optional[float] = None,
) -> bool:
    if amount is not None and transaction_type_he is not None:
        if is_refund_transaction(transaction_type_he, amount, charge_raw):
            return False
    if notes and "בקליטה" in notes:
        return True
    if charge_raw is None:
        return True
    try:
        return float(charge_raw) <= 0
    except (TypeError, ValueError):
        return True


def resolve_charge_ils(
    amount: float,
    charge_raw,
    merchant: str,
    notes: Optional[str],
    tx_date: Optional[date] = None,
    explicit_currency: Optional[str] = None,
    transaction_type_he: Optional[str] = None,
    transactions_for_matching: Optional[list] = None,
    skip_index: Optional[int] = None,
) -> tuple[float, Optional[str], bool]:
    """
    Return (charge_amount_ils, original_currency, estimated).
    Bank ILS charge (column 3) wins whenever present.
    """
    charge: Optional[float]
    if charge_raw is None:
        charge = None
    else:
        try:
            charge = float(charge_raw)
        except (TypeError, ValueError):
            charge = None

    if is_refund_transaction(transaction_type_he, amount, charge_raw):
        return resolve_refund_ils(
            amount,
            charge_raw,
            merchant,
            tx_date=tx_date,
            explicit_currency=explicit_currency,
            transaction_type_he=transaction_type_he,
            transactions_for_matching=transactions_for_matching,
            skip_index=skip_index,
        )

    if amount <= 0:
        return 0.0, None, False

    if is_interest_row(merchant) and charge is not None and charge > 0:
        return _round_money(charge), "ILS", False

    pending = is_pending_charge(
        charge_raw,
        notes,
        transaction_type_he=transaction_type_he,
        amount=amount,
    )

    # Pending — column 3 not final; use FX even if charge_amount was already estimated.
    if pending:
        currency = (
            explicit_currency
            if explicit_currency and explicit_currency != "ILS"
            else detect_currency(merchant, amount, None, None)
        )
        if currency == "ILS":
            return _round_money(amount), "ILS", True
        rate_date = tx_date or date.today()
        rate = get_rate_to_ils(currency, rate_date)
        return _round_money(amount * rate), currency, True

    # Bank ILS charge — always trust column 3 when the bank charged in ILS.
    if charge is not None and charge > 0:
        currency = detect_currency(merchant, amount, charge, explicit_currency)
        if currency == "ILS" or abs(amount - charge) < 0.02:
            return _round_money(charge), "ILS" if abs(amount - charge) < 0.02 else currency, False
        return _round_money(charge), currency, False

    currency = detect_currency(merchant, amount, charge, explicit_currency)
    if currency == "ILS":
        return _round_money(amount), "ILS", False

    rate_date = tx_date or date.today()
    rate = get_rate_to_ils(currency, rate_date)
    return _round_money(amount * rate), currency, True

File: api/src/utils/fx.ts
/** Foreign-currency → ILS for Leumi Visa exports (mirrors backend/fx.py). */
import type { Transaction } from "../types.js";
import { getRateToIls } from "./fxRates.js";

const HEBREW = /[\u0590-\u05FF]/;
const COUNTRY_SUFFIX = /\s([A-Z]{2})\s*$/;
const USD_MERCHANT =
  /OPENAI|CURSOR|AWS\s|APPLE\.COM|MIDJOURNEY|AIRBNB|GOOGLE|GITHUB|MICROSOFT|NETFLIX|SPOTIFY|TRIP\.COM|AIRALO|OMIO|BOUNCE|FILTERLY|SPECIALSCOMEDY|yollacalls|CONCERTPLACE|BANDCAMP|economybookings|TradeInn|Farfetch|HOTEL & RESORTS|DRIVALIA|SKIS ROSSIGNOL|EL AL \d/i;
const EUR_MERCHANT =
  /\b(SRL|SPA|OOD|EOOD|GMBH|TERMINI|ROMA|MILANO|AUTOGRILL|UBR\*|AVIS|ZARA|COS|UNIQLO|POLENE|VIvat|PAUL AIRPORT|DEDEM SPA|PIERLUIGI|SUMUP|MANGO ROMA)\b/i;
const BGN_MERCHANT = /\b(EOOD|OOD|BALGARIYA|BURGAS|BULGARIA|AMREST KOFI)\b/i;
const ILS_MERCHANT = /\bBIT\b/i;

function roundMoney(value: number): number {
  return Math.round(value * 100) / 100;
}

function hasHebrew(text: string): boolean {
  return HEBREW.test(text);
}

export function inferCurrencyFromRatio(amount: number, charge: number): string | null {
  if (amount <= 0 || charge <= 0) return null;
  const ratio = charge / amount;
  if (ratio >= 3.2 && ratio <= 4.05) return "USD";
  if (ratio >= 3.85 && ratio <= 4.55) return "EUR";
  if (ratio >= 1.85 && ratio <= 2.25) return "BGN";
  if (ratio >= 4.5 && ratio <= 5.0) return "GBP";
  if (ratio >= 0.85 && ratio <= 1.05) return "PLN";
  if (ratio >= 0.005 && ratio <= 0.02) return "AMD";
  if (ratio >= 0.98 && ratio <= 1.02) return "ILS";
  return null;
}

export function detectCurrency(
  merchant: string,
  amount: number,
  charge: number | null,
  explicitCurrency?: string | null,
): string {
  if (explicitCurrency && explicitCurrency !== "ILS") return explicitCurrency;

  if (hasHebrew(merchant)) return "ILS";

  if (ILS_MERCHANT.test(merchant)) return "ILS";

  const suffix = COUNTRY_SUFFIX.exec(merchant.trim());
  if (suffix && suffix[1] === "IL") return "ILS";

  if (charge != null && charge > 0 && amount > 0) {
    const inferred = inferCurrencyFromRatio(amount, charge);
    if (inferred) return inferred;
  }

  if (USD_MERCHANT.test(merchant)) return "USD";

  if (suffix) {
    const cc = suffix[1]!;
    if (cc === "IL") return "ILS";
    if (cc === "US") return "USD";
    if (cc === "PL") return "PLN";
    if (["DE", "FR", "IT", "ES", "NL", "AT", "BE", "PT", "GR"].includes(cc)) return "EUR";
    if (cc === "GB") return "GBP";
    if (cc === "BG") return "BGN";
    if (cc === "AM") return "AMD";
  }

  if (BGN_MERCHANT.test(merchant)) return "BGN";
  if (EUR_MERCHANT.test(merchant)) return "EUR";
  if (/[A-Za-z]{3,}/.test(merchant) && !hasHebrew(merchant)) return "USD";

  return "ILS";
}

export function pendingCurrencyForAmount(
  amount: number,
  pendingCurrencies?: Record<string, string> | null,
): string | undefined {
  if (!pendingCurrencies) return undefined;
  const key = roundMoney(amount);
  const keys = [String(key), key.toFixed(2), key.toFixed(1)];
  for (const k of keys) {
    const hit = pendingCurrencies[k];
    if (hit) return hit;
  }
  return undefined;
}

function explicitCurrencyForTx(
  tx: Transaction,
  pendingCurrencies?: Record<string, string> | null,
): string | null | undefined {
  if (isPendingCharge(tx.charge_amount, tx.notes, tx.transaction_type_he, tx.amount)) {
    // Header map and parser output beat merchant heuristics (OPENAI → USD).
    return (
      pendingCurrencyForAmount(tx.amount, pendingCurrencies) ??
      tx.original_currency ??
      null
    );
  }
  return tx.original_currency;
}

const REFUND_TYPE_MARKERS = ["זיכוי", "השבת מכירה"] as const;
const INSTALLMENT = /תשלום\s+(\d+)\s+מתוך\s+(\d+)/;
const BILLING_ADJUSTMENT = "שינוי מועד חיוב";
const CARD_FEE_TYPE = "דמי חבר";
const INTEREST_MERCHANT = "* ריבית *";

export function parseInstallment(notes: string | null | undefined): [number, number] | null {
  if (!notes) return null;
  const match = INSTALLMENT.exec(notes);
  if (!match) return null;
  return [Number(match[1]), Number(match[2])];
}

export function shouldSkipInstallmentRow(
  chargeRaw: number | null | undefined,
  notes: string | null,
  transactionTypeHe?: string | null,
): boolean {
  const installment = parseInstallment(notes);
  if (!installment) return false;
  if (installment[0] === 0) return true;
  return isPendingCharge(chargeRaw, notes, transactionTypeHe);
}

export function isBillingCycleAdjustment(merchant: string): boolean {
  return merchant.includes(BILLING_ADJUSTMENT);
}

export function isCardMembershipFee(merchant: string, transactionTypeHe?: string | null): boolean {
  if (transactionTypeHe === CARD_FEE_TYPE) return true;
  return merchant.includes("דמי כרטיס");
}

export function isInterestRow(merchant: string): boolean {
  return merchant.includes(INTEREST_MERCHANT);
}

export function shouldSkipNonSpendRow(
  chargeRaw: number | null | undefined,
  merchant: string,
  notes: string | null,
  transactionTypeHe?: string | null,
): boolean {
  if (shouldSkipInstallmentRow(chargeRaw, notes, transactionTypeHe)) return true;
  if (isBillingCycleAdjustment(merchant)) return true;
  if (isCardMembershipFee(merchant, transactionTypeHe) && isPendingCharge(chargeRaw, notes, transactionTypeHe)) {
    return true;
  }
  return false;
}

export function transactionSnapshotKey(tx: Transaction): string {
  const txDate = tx.date.slice(0, 10);
  const installment = parseInstallment(tx.notes);
  if (installment) {
    return `${txDate}|${tx.merchant_he}|inst|${installment[0]}|${installment[1]}`;
  }
  const charge = roundMoney(tx.charge_amount);
  return `${txDate}|${tx.merchant_he}|tx|${charge}|${tx.notes ?? ""}`;
}

export function pickBetterTransaction(a: Transaction, b: Transaction): Transaction {
  const aEst = !!a.charge_estimated;
  const bEst = !!b.charge_estimated;
  if (aEst !== bEst) return aEst ? b : a;

  const aInst = parseInstallment(a.notes);
  const bInst = parseInstallment(b.notes);
  if (aInst && bInst && aInst[0] !== bInst[0]) {
    return aInst[0] > bInst[0] ? a : b;
  }

  if (a.charge_amount !== b.charge_amount) {
    return a.charge_amount < b.charge_amount ? a : b;
  }

  return b;
}

export function dedupeTransactionSnapshots(transactions: Transaction[]): Transaction[] {
  const best = new Map<string, Transaction>();
  const order: string[] = [];
  for (const tx of transactions) {
    const key = transactionSnapshotKey(tx);
    if (!best.has(key)) {
      order.push(key);
      best.set(key, tx);
    } else {
      best.set(key, pickBetterTransaction(best.get(key)!, tx));
    }
  }
  return order.map((key) => best.get(key)!);
}

export function isRefundTypeHe(transactionTypeHe: string | null | undefined): boolean {
  if (!transactionTypeHe) return false;
  return REFUND_TYPE_MARKERS.some((m) => transactionTypeHe.includes(m));
}

export function isRefundTransaction(
  transactionTypeHe: string | null | undefined,
  amount: number,
  chargeRaw: number | null | undefined,
): boolean {
  if (isRefundTypeHe(transactionTypeHe)) return true;
  if (amount < 0) return true;
  if (chargeRaw != null && chargeRaw < 0) return true;
  return false;
}

function txDateValue(tx: Transaction): string {
  return tx.date.slice(0, 10);
}

export function findMatchingOriginalCharge(
  merchant: string,
  foreignAmount: number,
  refundDate: string,
  transactions: Transaction[],
  skipIndex?: number,
): Transaction | null {
  if (foreignAmount <= 0) return null;
  let best: Transaction | null = null;
  let bestDate = "";
  for (let i = 0; i < transactions.length; i++) {
    if (skipIndex != null && i === skipIndex) continue;
    const tx = transactions[i]!;
    if (tx.merchant_he !== merchant) continue;
    if (isRefundTransaction(tx.transaction_type_he, tx.amount, tx.charge_amount)) continue;
    if (tx.charge_amount <= 0) continue;
    if (Math.abs(Math.abs(tx.amount) - foreignAmount) > 0.02) continue;
    const txDate = txDateValue(tx);
    if (txDate > refundDate) continue;
    if (!best || txDate > bestDate) {
      best = tx;
      bestDate = txDate;
    }
  }
  return best;
}

export function resolveRefundIls(
  amount: number,
  chargeRaw: number | null | undefined,
  merchant: string,
  txDate: string,
  explicitCurrency?: string | null,
  transactionsForMatching?: Transaction[] | null,
  skipIndex?: number,
): { chargeAmount: number; originalCurrency: string | null; estimated: boolean } {
  const charge = chargeRaw == null ? null : Number(chargeRaw);
  const foreignAmount = amount !== 0 ? Math.abs(amount) : charge != null && charge !== 0 ? Math.abs(charge) : 0;
  const refundDate = txDate.slice(0, 10);

  if (transactionsForMatching?.length) {
    const match = findMatchingOriginalCharge(
      merchant,
      foreignAmount,
      refundDate,
      transactionsForMatching,
      skipIndex,
    );
    if (match) {
      return {
        chargeAmount: roundMoney(-Math.abs(match.charge_amount)),
        originalCurrency: match.original_currency ?? "ILS",
        estimated: !!match.charge_estimated,
      };
    }
  }

  if (charge != null && charge !== 0 && foreignAmount > 0) {
    const ilsMagnitude = Math.abs(charge);
    if (Math.abs(ilsMagnitude - foreignAmount) > 0.02) {
      return { chargeAmount: roundMoney(-ilsMagnitude), originalCurrency: "ILS", estimated: false };
    }
  }

  if (explicitCurrency && explicitCurrency !== "ILS") {
    const rate = getRateToIls(explicitCurrency, txDate);
    return {
      chargeAmount: roundMoney(-foreignAmount * rate),
      originalCurrency: explicitCurrency,
      estimated: true,
    };
  }

  const base = charge != null && charge !== 0 ? Math.abs(charge) : foreignAmount;
  return { chargeAmount: roundMoney(-base), originalCurrency: "ILS", estimated: false };
}

export function isPendingCharge(
  chargeRaw: number | null | undefined,
  notes: string | null,
  transactionTypeHe?: string | null,
  amount?: number,
): boolean {
  if (amount != null && transactionTypeHe != null) {
    if (isRefundTransaction(transactionTypeHe, amount, chargeRaw)) return false;
  }
  if (notes?.includes("בקליטה")) return true;
  if (chargeRaw == null) return true;
  return chargeRaw <= 0;
}

export function resolveChargeIls(
  amount: number,
  chargeRaw: number | null | undefined,
  merchant: string,
  notes: string | null,
  txDate: string,
  explicitCurrency?: string | null,
  transactionTypeHe?: string | null,
  transactionsForMatching?: Transaction[] | null,
  skipIndex?: number,
): { chargeAmount: number; originalCurrency: string | null; estimated: boolean } {
  const charge = chargeRaw == null ? null : Number(chargeRaw);

  if (isRefundTransaction(transactionTypeHe, amount, chargeRaw)) {
    return resolveRefundIls(
      amount,
      chargeRaw,
      merchant,
      txDate,
      explicitCurrency,
      transactionsForMatching,
      skipIndex,
    );
  }

  if (amount <= 0) return { chargeAmount: 0, originalCurrency: null, estimated: false };

  if (isInterestRow(merchant) && charge != null && charge > 0) {
    return { chargeAmount: roundMoney(charge), originalCurrency: "ILS", estimated: false };
  }

  const pending = isPendingCharge(chargeRaw, notes, transactionTypeHe, amount);

  // Pending — column 3 not final; use FX even if charge_amount was already estimated.
  if (pending) {
    const currency =
      explicitCurrency && explicitCurrency !== "ILS"
        ? explicitCurrency
        : detectCurrency(merchant, amount, null, null);
    if (currency === "ILS") {
      return { chargeAmount: roundMoney(amount), originalCurrency: "ILS", estimated: true };
    }
    const rate = getRateToIls(currency, txDate);
    return {
      chargeAmount: roundMoney(amount * rate),
      originalCurrency: currency,
      estimated: true,
    };
  }

  // Bank ILS charge — column 3 always wins.
  if (charge != null && charge > 0) {
    const currency = detectCurrency(merchant, amount, charge, explicitCurrency);
    if (currency === "ILS" || Math.abs(amount - charge) < 0.02) {
      return {
        chargeAmount: roundMoney(charge),
        originalCurrency: Math.abs(amount - charge) < 0.02 ? "ILS" : currency,
        estimated: false,
      };
    }
    return { chargeAmount: roundMoney(charge), originalCurrency: currency, estimated: false };
  }

  const currency = detectCurrency(merchant, amount, charge, explicitCurrency);
  if (currency === "ILS") {
    return { chargeAmount: roundMoney(amount), originalCurrency: "ILS", estimated: false };
  }

  const rate = getRateToIls(currency, txDate);
  return {
    chargeAmount: roundMoney(amount * rate),
    originalCurrency: currency,
    estimated: true,
  };
}

export function normalizeForeignCharges(
  transactions: Transaction[],
  pendingCurrencies?: Record<string, string> | null,
): {
  transactions: Transaction[];
  changed: boolean;
} {
  let changed = false;
  const result = transactions.map((tx, index) => {
    const resolved = isRefundTransaction(tx.transaction_type_he, tx.amount, tx.charge_amount)
      ? resolveRefundIls(
          tx.amount,
          tx.charge_amount,
          tx.merchant_he,
          tx.date,
          explicitCurrencyForTx(tx, pendingCurrencies),
          transactions,
          index,
        )
      : resolveChargeIls(
          tx.amount,
          tx.charge_amount,
          tx.merchant_he,
          tx.notes,
          tx.date,
          explicitCurrencyForTx(tx, pendingCurrencies),
          tx.transaction_type_he,
        );
    const sameCharge = resolved.chargeAmount === tx.charge_amount;
    const sameCurrency = (tx.original_currency ?? null) === resolved.originalCurrency;
    const sameEstimated = !!tx.charge_estimated === resolved.estimated;
    if (sameCharge && sameCurrency && sameEstimated) return tx;

    changed = true;
    return {
      ...tx,
      charge_amount: resolved.chargeAmount,
      original_currency: resolved.originalCurrency ?? undefined,
      charge_estimated: resolved.estimated,
    };
  });
  return { transactions: result, changed };
}
